test/functional/mining.py

```diff
@@ -0,0 +1,336 @@
+        #   block we've seen so far.
+
+        # First, convert recent_tx_window from milliseconds to seconds.
+        recent_tx_window = int((recent_tx_window+999)/1000)
+
+        mempool = node.getrawmempool(verbose=True)
+        all_times = [ value["time"] for key, value in mempool.items() ]
+        earliest_time, latest_time = sorted(all_times)[::len(all_times)-1]
+        mock_time = latest_time + recent_tx_window - 1
+        max_fee = 0
+
+        # Track how many blocks are found with/without recent transactions
+        blocks_with_recent_tx = 0
+        blocks_without_recent_tx = 0
+
+        while mock_time >= earliest_time + recent_tx_window:
+            node.setmocktime(mock_time)
+            self.add_empty_block(node) # bypass gbt cache
+            template = node.getblocktemplate({"rules":["segwit"]})
+            template_txids = [ x['txid'] for x in template['transactions'] ]
+
+            block_fee = 0
+            consensus_fee = 0
+            contains_recent_tx = False
+            recent_tx_weight = 0
+            for x in template['transactions']:
+                block_fee += int(mempool[x['txid']]['modifiedfee']*COIN)
+                consensus_fee += x['fee']
+                if (mempool[x['txid']]['time'] > mock_time - recent_tx_window):
+                    self.log.debug("Contains recent tx %s time %d (%d, %d)",
+                            x['txid'], mempool[x['txid']]['time'], mock_time,
+                            recent_tx_window)
+                    contains_recent_tx = True
+                    recent_tx_weight += x['weight']
+            if block_fee > max_fee:
+                self.log.debug("Setting max_fee to %d", block_fee)
+                max_fee = block_fee
+            block_reward = template['coinbasevalue'] - consensus_fee
+
+            if (not contains_recent_tx):
+                # If we don't have anything new, we better be close to the max
+                # seen so far
+                self.log.debug("No recent transactions, block_fee = %d, block_reward=%d", block_fee, block_reward)
+                assert (block_reward + block_fee >= (1-recent_tx_stale_rate)*(block_reward + max_fee))
+                blocks_without_recent_tx += 1
+
+            else:
+                # Deprioritise everything recent and re-run; the fee delta
+                # should be significant
+                blocks_with_recent_tx += 1
+
+                for txid, entry in mempool.items():
+                    if (entry["time"] > mock_time - recent_tx_window):
+                        node.prioritisetransaction(txid, int(-(entry["fee"]*COIN)))
+                        if (txid in template_txids):
+                            # The mining algorithm will include non-recent
+                            # ancestors of recently included transactions.
+                            # If this is improved in the future, this portion
+                            # of the test would need to change.
```

Comment:
In commit "Add functional test for CreateNewBlock"

Maybe spell this out a little more. It seems like the point of setting fees to 0 is to force gbt to generate and return the `noRecent` block which lost the competition against the recent block. But if the `noRecent` block currently includes these ancestor transactions, then why set their fees to 0 here?